src/lib/utils/bitvector.h

```diff
@@ -0,0 +1,1327 @@
+      }
+
+      /// @name Serialization
+      /// @{
+
+      /**
+       * Re-initialize the bitvector with the given bytes. See the respective
+       * constructor for details. This should be used only when trying to save
+       * allocations. Otherwise, use the constructor.
+       *
+       * @param bytes  the byte range to load bits from
+       * @param bits   (optional) if not all @p bytes should be loaded in full
+       */
+      void from_bytes(std::span<const uint8_t> bytes, std::optional<size_type> bits = std::nullopt) {
+         m_bits = bits.value_or(bytes.size_bytes() * 8);
+         BOTAN_ARG_CHECK(m_bits <= bytes.size_bytes() * 8, "not enough data to load so many bits");
+         resize(m_bits);
+
+         // load as much aligned data as possible
+         const auto verbatim_blocks = m_bits / block_size_bits;
+         const auto verbatim_bytes = verbatim_blocks * block_size_bytes;
+         if(verbatim_blocks > 0) {
+            typecast_copy(std::span{m_blocks}.first(verbatim_blocks), bytes.first(verbatim_bytes));
+         }
+
+         // load remaining unaligned data
+         for(size_type i = verbatim_bytes * 8; i < m_bits; ++i) {
+            ref(i) = ((bytes[i >> 3] & (uint8_t(1) << (i & 7))) != 0);
+         }
+      }
+
+      /**
+       * Renders the bitvector into a byte array. By default, this will use
+       * `std::vector<uint8_t>` or `Botan::secure_vector<uint8_t>`, depending on
+       * the allocator used by the bitvector. The rendering is compatible with
+       * the bit layout explained in the respective constructor.
+       */
+      template <concepts::resizable_byte_buffer OutT =
+                   std::conditional_t<uses_secure_allocator, secure_vector<uint8_t>, std::vector<uint8_t>>>
+      OutT to_bytes() const {
+         OutT out(ceil_tobytes(m_bits));
+         to_bytes(out);
+         return out;
+      }
+
+      /**
+       * Renders the bitvector into a properly sized byte range.
+       *
+       * @param out  a byte range that has a length of at least `ceil_tobytes(size())`.
+       */
+      void to_bytes(std::span<uint8_t> out) const {
+         const auto bytes_needed = ceil_tobytes(m_bits);
+         BOTAN_ARG_CHECK(bytes_needed <= out.size_bytes(), "Not enough space to render bitvector");
+
+         // copy as much aligned data as possible
+         const auto verbatim_blocks = m_bits / block_size_bits;
+         const auto verbatim_bytes = verbatim_blocks * block_size_bytes;
+         if(verbatim_blocks > 0) {
+            typecast_copy(out.first(verbatim_bytes), std::span{m_blocks}.first(verbatim_blocks));
+         }
+
+         // copy remaining unaligned data
+         for(size_type i = verbatim_bytes * 8; i < m_bits; ++i) {
+            out[i >> 3] |= ref(i).template as<uint8_t>() << (i & 7);
+         }
+      }
+
+      /**
+       * Renders this bitvector into a sequence of "0"s and "1"s.
+       */
+      std::string to_string() const {
```

Comment:
When would this be useful? If just for tests/debugging maybe better to define it elsewhere

Reply:
Fine to leave it here